server/src/main/java/org/elasticsearch/usage/UsageService.java

```diff
@@ -89,6 +92,33 @@ public void addRestHandler(BaseRestHandler handler) {
         }
     }
 
+    public static final String OTHER_SUBTYPE = "other";
+
+    public void registerAggregationUsage(String aggregationName) {
+        registerAggregationUsage(aggregationName, OTHER_SUBTYPE);
+    }
+
+    public void registerAggregationUsage(String aggregationName, String valuesSourceType) {
+        Map<String, LongAdder> subAgg = aggs.computeIfAbsent(aggregationName, k -> new HashMap<>());
+        if ( subAgg.put(valuesSourceType, new LongAdder()) != null) {
+            throw new IllegalArgumentException("stats for aggregation [" + aggregationName + "][" + valuesSourceType +
+                "] already registered");
+        }
+    }
+
+    public void incAggregationUsage(String aggregationName, String valuesSourceType) {
+        Map<String, LongAdder> valuesSourceMap = aggs.get(aggregationName);
+        // Not all aggs register their usage at the moment we also don't register them in test context
+        if (valuesSourceMap != null) {
+            LongAdder adder = valuesSourceMap.get(valuesSourceType);
+            if (adder != null) {
+                adder.increment();
+            } else {
+                throw new IllegalArgumentException("Unknown subtype [" + aggregationName + "][" + valuesSourceType + "]");
```

Comment:
I wonder if we should just log an error but not actually throw an exception?  I'm thinking if we end up breaking something on the usage side, it'd be a shame if it actually broke the aggs until we got a bugfix out

Reply:
That helped me to find a lot of strange behaving aggregations. But you are right, I might just replace this with assert for now.